apps/editing-toolkit/editing-toolkit-plugin/tutorials/class-wpcom-tutorials.php

```diff
@@ -0,0 +1,253 @@
+ * a `WpcomTourKit` component on the frontend, but we are only concerned with the top
+ * two levels of Tutorial and its contained Tasks, tracking Task completion status
+ * on the backend. This API will be used to generate
+ *
+ * @package A8C\FSE
+ */
+
+/**
+ * WPCom_Tutorials class
+ */
+class WPCom_Tutorials {
+
+	/**
+	 * Storage for our singleton instance.
+	 *
+	 * @var null|WPCom_Tutorials
+	 */
+	private static $instance = null;
+
+	/**
+	 * Internal storage for registered Tutorials.
+	 *
+	 * @var array
+	 */
+	private $registry = array();
+
+	/**
+	 * Gets our singleton instance, initiating it the first time.
+	 *
+	 * @return WPCom_Tutorials instance.
+	 */
+	public static function get_instance() {
+		if ( is_null( self::$instance ) ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Registers a Tutorial.
+	 *
+	 * @param string $id      A unique ID for the Tutorial.
+	 * @param array  $options `title` and `tasks` required.
+	 *
+	 * @return bool           True if successfully registered, false if it failed.
+	 */
+	public function register( $id, $options ) {
+		if ( ! $this->validate_options( $options ) ) {
+			return false;
+		}
+		$options['id']         = $id;
+		$this->registry[ $id ] = $options;
+		return true;
+	}
+
+	/**
+	 * Unregisters a registered Tutorial
+	 *
+	 * @param string $id Tutorial ID.
+	 * @return bool      True if successfully unregistered, false if not found.
+	 */
+	public function unregister( $id ) {
+		if ( ! isset( $this->registry[ $id ] ) ) {
+			return false;
+		}
+		unset( $this->registry[ $id ] );
+		return true;
+	}
+
+	/**
+	 * Validates the options passed in `WPCom_Tutorials::register`.
+	 *
+	 * @param array $options Associative array passed with registration.
+	 * @return bool          Validation success.
+	 */
+	private function validate_options( $options ) {
+		if ( ! isset( $options['title'] ) || ! is_string( $options['title'] ) ) {
+			return false;
+		}
+		if ( ! isset( $options['tasks'] ) || ! is_array( $options['tasks'] ) ) {
+			return false;
+		}
+		foreach ( $options['tasks'] as $task ) {
+			if ( ! $this->validate_task( $task ) ) {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * Validates a single task for expected structure.
+	 *
+	 * @param array $task A single task (id and title).
+	 * @return bool       Validation success.
+	 */
+	private function validate_task( $task ) {
+		return isset( $task['id'] )
+			&& is_string( $task['id'] )
+			&& isset( $task['title'] )
+			&& is_string( $task['title'] );
+	}
+
+	/**
+	 * Retrieves a Tutorial from the registry.
+	 *
+	 * @param string $id  The ID of the Tutorial to fetch.
+	 * @return null|array Tutorial associative array if found, null if not.
+	 */
+	public function get_tutorial( $id ) {
+		if ( ! isset( $this->registry[ $id ] ) ) {
+			return null;
+		}
+		$tutorial = $this->registry[ $id ];
+
+		// Populating status at runtime so that any invalidations happen within the Options API.
```

Comment:
This is me hoping to learn more. Would you mind explaining this comment more? What are the advantages of having invalidations happen within the Options API?

Reply:
Ah good find! This should read "User (meta/attributes) API" since I transitioned an early draft of this API to store data on a per-site basis with options, but we quickly realized this should be keyed to users instead, so we migrated to user attributes (dotcom) or meta (dotorg).

The bit about invalidations means that by leveraging a Core API for storing our state, we don't have to worry about the current truth at any given point, whenever a tutorial is requested, its currently stored progress will be consistent. Underlying all of that is the DB writes, memcached invalidations, datacenter sync, etc, that means we're always using the latest data by deferring to those APIs rather than trying to make an explicit table for tutorial steps with a structure catering to that. Doing things inside WP is more about finessing existing APIs to your needs rather than being able to do something more obvious and direct.

Reply:
> Doing things inside WP is more about finessing existing APIs to your needs rather than being able to do something more obvious and direct.

Lots of wisdom here 👍 

That makes sense! Ty ty